File: src/main/java/hu/frontrider/arcana/items/Formula.java
package hu.frontrider.arcana.items;

import hu.frontrider.arcana.ThaumicArcana;
import hu.frontrider.arcana.creatureenchant.backend.CreatureEnchant;
import hu.frontrider.arcana.items.formula.BaseFormula;
import hu.frontrider.arcana.recipes.formulacrafting.FormulaRecipes;
import net.minecraft.client.util.ITooltipFlag;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.item.ItemStack;
import net.minecraft.util.NonNullList;
import net.minecraft.util.ResourceLocation;
import net.minecraft.world.World;
import net.minecraftforge.fml.common.registry.GameRegistry;
import net.minecraftforge.registries.IForgeRegistry;

import javax.annotation.Nullable;
import java.util.List;
import java.util.stream.Collectors;

public class Formula extends BaseFormula {
    private IForgeRegistry<CreatureEnchant> creatureEnchantIForgeRegistry;

    public Formula() {
        super(new ResourceLocation(ThaumicArcana.MODID, "formula"));
        setMaxStackSize(1);
        creatureEnchantIForgeRegistry = GameRegistry.findRegistry(CreatureEnchant.class);
    }

    @Override
    public void addInformation(ItemStack stack, @Nullable World worldIn, List<String> tooltip, ITooltipFlag flagIn) {
        super.addInformation(stack, worldIn, tooltip, flagIn);
    }

    @Override
    public void getSubItems(CreativeTabs tab, NonNullList<ItemStack> items) {
        if (tab != ThaumicArcana.INSTANCE.getTABARCANA())
            return;

        items.add(new ItemStack(this));
        items.addAll(FormulaRecipes.INSTANCE.getRecipesCollection()
                .stream()
                .flatMap(List::stream)
                .map(recipe -> getItemStack(recipe.getFormula()))
                .collect(Collectors.toSet()));
    }
}

File: src/main/java/hu/frontrider/arcana/content/items/formula/Formula.kt
package hu.frontrider.arcana.content.items.formula

import hu.frontrider.arcana.ThaumicArcana
import hu.frontrider.arcana.content.items.formula.BaseFormula
import hu.frontrider.arcana.core.creatureenchant.CreatureEnchant
import hu.frontrider.arcana.core.formulacrafting.FormulaApplicationRecipe

import hu.frontrider.arcana.core.formulacrafting.FormulaRecipes
import net.minecraft.client.util.ITooltipFlag
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.item.ItemStack
import net.minecraft.util.NonNullList
import net.minecraft.util.ResourceLocation
import net.minecraft.world.World
import net.minecraftforge.fml.common.registry.GameRegistry
import net.minecraftforge.registries.IForgeRegistry
import java.util.stream.Collector
import java.util.stream.Collectors
import java.util.stream.Stream

class Formula : BaseFormula(ResourceLocation(ThaumicArcana.MODID, "formula")) {
    private val creatureEnchantIForgeRegistry: IForgeRegistry<CreatureEnchant>

    init {
        setMaxStackSize(1)
        creatureEnchantIForgeRegistry = GameRegistry.findRegistry(CreatureEnchant::class.java)
    }

    override fun addInformation(stack: ItemStack, worldIn: World?, tooltip: MutableList<String>, flagIn: ITooltipFlag) {
        super.addInformation(stack, worldIn, tooltip, flagIn)
    }

    override fun getSubItems(tab: CreativeTabs, items: NonNullList<ItemStack>) {
        if (tab !== ThaumicArcana.TABARCANA)
            return

        items.add(ItemStack(this))
        items.addAll(FormulaRecipes.getRecipesCollection()
                .stream()
                .flatMap<FormulaApplicationRecipe> { it.stream() }
                .map { recipe -> getItemStack(recipe.formula) }
                .collect(Collectors.toSet()))
    }
}
